### PR TITLE
#595 Implemented Grouped class for grouping iterable elements

### PR DESCRIPTION
**Pull request by nick318**
#595 Implemented class `Grouped` for grouping elements of iterable object.
Usage:
```
new Grouped<>(
       new IterableOf<>(1, 1, 1, 4, 5, 6, 7, 8, 9),
              number -> number,
              Object::toString
        )
```

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#647](https://github.com/yegor256/cactoos/issues/647) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#647](https://github.com/yegor256/cactoos/issues/647) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C63314D6Z/452023816). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by nick318** [comment]
@amihaiemil any thoughts about PR?)

File: src/main/java/org/cactoos/map/Grouped.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * functions for keys and values</p>
+ *
+ * <p>There is no thread-safety guarantee.
+ *
+ * @author Nikita Salomatin (nsalomatin@hotmail.com)
+ * @version $Id$
+ * @param <T> Type of key
+ * @param <V> Type of value
+ * @param <K> Type of entry objects of functions
+ * @since 0.4
```
Comment (amihaiemil) [inline src/main/java/org/cactoos/map/Grouped.java]: @nick318 ``@since 0.30`` -- always look on the repo's Releases page. Now the last release is 0.29, so we assume the next one will be 0.30

File: src/main/java/org/cactoos/map/Grouped.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * Iterable as {@link Map}.
+ *
+ * <p>This class groups objects from iterable by applying
+ * functions for keys and values</p>
+ *
+ * <p>There is no thread-safety guarantee.
+ *
+ * @author Nikita Salomatin (nsalomatin@hotmail.com)
+ * @version $Id$
+ * @param <T> Type of key
```
Comment (amihaiemil) [inline src/main/java/org/cactoos/map/Grouped.java]: @nick318 Are you sure these are in the right order? ``T`` is for type of key? Seems strange, I would say ``K`` should be the type of the key

File: src/main/java/org/cactoos/map/Grouped.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+public final class Grouped<T, V, K> extends MapEnvelope<T, List<V>> {
+
+    /**
+     * Ctor.
+     *
+     * @param list Iterable which is used to retrieve data from
+     * @param keys Function to get a key
+     * @param values Function to get a value
+     */
+    public Grouped(
```
Comment (amihaiemil) [inline src/main/java/org/cactoos/map/Grouped.java]: @nick318 Indentation here is a bit off, how about this? 

```java
public Grouped(
    final Iterable<K> list,
    final Function<K, T> keys,
    final Function<K, V> values
) {
    super(
        () -> {
            final Stream<K> stream = StreamSupport.stream(
                list.spliterator(), false
            );
            return stream.collect(
                Collectors.groupingBy(
                    keys,
                    Collectors.mapping(values, Collectors.toList())
                )
            );
        }
    );
}
```

So, 4-space indentation + every closing bracket on its own line and lines no longer than 80 chars.

File: src/test/java/org/cactoos/map/GroupedTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ * @author Nikita Salomatin (nsalomatin@hotmail.com)
+ * @version $Id$
+ * @since 0.4
+ * @checkstyle JavadocMethodCheck (500 lines)
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+public final class GroupedTest {
+
+    @Test
+    public void groupedByNumber() throws Exception {
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/map/GroupedTest.java]: @nick318 can you add a few more tests? One for empty iterables and one that would better show what exactly this class does? It is not so clear from thsi unit test alone.

File: src/main/java/org/cactoos/map/Grouped.java (hunk 1 above)
Comment (nick318) [reply src/main/java/org/cactoos/map/Grouped.java]: @amihaiemil done

File: src/main/java/org/cactoos/map/Grouped.java (hunk 2 above)
Comment (nick318) [reply src/main/java/org/cactoos/map/Grouped.java]: @amihaiemil done

File: src/main/java/org/cactoos/map/Grouped.java (hunk 3 above)
Comment (nick318) [reply src/main/java/org/cactoos/map/Grouped.java]: @amihaiemil done

File: src/test/java/org/cactoos/map/GroupedTest.java (hunk 4 above)
Comment (nick318) [reply src/test/java/org/cactoos/map/GroupedTest.java]: @amihaiemil done

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job [#647](https://github.com/yegor256/cactoos/issues/647) is already in scope

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) resigned from [#647](https://github.com/yegor256/cactoos/issues/647), since the job is not in scope anymore

**Comment by 0crat** [comment]
The job [#647](https://github.com/yegor256/cactoos/issues/647) is now out of scope
